components/nexus-repository/src/main/java/org/sonatype/nexus/repository/util/SimpleStateGuard.java

```diff
@@ -0,0 +1,262 @@
+    finally {
+      unlock();
+    }
+  }
+
+  /**
+   * Ensure the current state is one of allowed.
+   */
+  public void ensure(final T... allowed) {
+    checkNotNull(allowed);
+    checkArgument(allowed.length != 0);
+
+    lock();
+    try {
+      _ensure(allowed);
+    }
+    finally {
+      unlock();
+    }
+  }
+
+  private void _ensure(final T[] allowed) {
+    for (T allow : allowed) {
+      if (current == allow) {
+        return;
+      }
+    }
+
+    throw new IllegalStateException("Invalid state: " + current.name() + "; allowed: " + Arrays.toString(allowed));
+  }
+
+  /**
+   * Create a transition to given state.
+   */
+  public Transition transition(final T to) {
+    return new Transition(to);
+  }
+
+  /**
+   * Create a guard which allows execution in the given states.
+   */
+  public Guard guard(final T... allowed) {
+    return new Guard(allowed);
+  }
+
+  //
+  // Transition
+  //
+
+  /**
+   * Transition from current state to target state and execute an action.
+   */
+  public class Transition
+  {
+    private final T to;
+
+    @Nullable
+    private T[] allowed;
+
+    public Transition(final T to) {
+      this.to = checkNotNull(to);
+    }
+
+    public Transition from(final T... allowed) {
+      checkNotNull(allowed);
+      checkArgument(allowed.length != 0);
+      this.allowed = allowed;
+      return this;
+    }
+
+    public void run(final Action action) throws Exception {
+      lock();
+      try {
+        if (allowed != null) {
+          _ensure(allowed);
+        }
+
+        try {
+          log.trace("Transitioning: {} -> {}", current, to);
+
+          action.run();
+          current = to;
+
+          log.trace("Transitioned: {}", to);
+        }
+        catch (Throwable t) {
+          log.error("Failed transition: {} -> {}", current, to, t);
+
+          // maybe set failure state
+          if (failure != null) {
+            current = failure;
+          }
+
+          Throwables.propagateIfPossible(t, Exception.class, Error.class);
+          throw Throwables.propagate(t);
+        }
+      }
+      finally {
+        unlock();
+      }
+    }
+  }
+
+  //
+  // Guard
+  //
+
+  /**
+   * Execute an action if current state is allowed.
+   */
+  public class Guard
+  {
+    private final T[] allowed;
+
+    public Guard(final T[] allowed) {
+      checkNotNull(allowed);
+      checkArgument(allowed.length != 0);
+      this.allowed = allowed;
+    }
+
+    public void run(final Action action) throws Exception {
+      lock();
+      try {
+        _ensure(allowed);
+        action.run();
+      }
+      finally {
+        unlock();
+      }
+    }
+  }
+
+  //
+  // Action
+  //
+
+  public static interface Action
+  {
+    void run() throws Exception;
+  }
+
+  //
+  // Builder
+  //
+
+  public static class Builder<T extends Enum>
+  {
+    private Lock lock;
+
+    private T initial;
+
+    private T failure;
+
+    // final states?
+
+    public Builder<T> lock(final Lock lock) {
```

Comment:
Yup, I originally had the builder supply a lock if none was given but went back to requiring user to do that, find to put that back.